bootstrap.sh

```diff
@@ -47,109 +59,104 @@ if [ "$HOME" != "/root" ]; then
+        shift; if [ -z "${1}" ]; then usage; fi
+        AGENTS_BRANCH="${1}"; BRANCH_PARAM_SPECIFIED="Y"
+        shift; continue
     fi
-    if [ "${1}" = "--force-upgrade" ]; then
-        FORCE_UPGRADE="Y"
-        shift
-        continue
+    if [ "${1}" = "--thirdparties-repo" ]; then
+        shift; if [ -z "${1}" ]; then usage; fi
+        THIRDPARTIES_REPO="${1}"; BRANCH_PARAM_SPECIFIED="Y"
+        shift; continue
+    fi
+    if [ "${1}" = "--thirdparties-branch" ]; then
+        shift; if [ -z "${1}" ]; then usage; fi
+        THIRDPARTIES_BRANCH="${1}"; BRANCH_PARAM_SPECIFIED="Y"
+        shift; continue
+    fi
+    if [ "${1}" = "--tag" ]; then
+        shift; if [ -z "${1}" ]; then usage; fi
+        GIT_TAG="${1}";
+        if [${GIT_TAG} != "dev" ]; then TAG_PARAM_SPECIFIED="Y"; fi
+        shift; continue
     fi
     if [ "${1}" = "-t" ]; then
         UNIT_TEST="Y"
-        shift
-        continue
+        shift; continue
     fi
+    echo
+    echo "Parameter '$1' is not recognized"
+    echo
     usage
 done
 
-# Empty tag means we use the latest available.
-if [ -z "${GIT_TAG}" ]; then
-    GIT_TAG="last"
-else
-    # We don't use Tag and Branch at the same time.
-    if [ "${FORCE_UPGRADE}" = "N"  ] && [  "${APPSCALE_BRANCH}" != "master" ]; then
-        echo "--branch cannot be specified with --tag"
-        exit 1
-    fi
+
+# Validate parameters combination
+if [ "${BRANCH_PARAM_SPECIFIED}" = "Y" ] && [ "${TAG_PARAM_SPECIFIED}" = "Y" ]; then
+    echo "Repo/Branch parameters can't be used if --tag parameter is specified"
+    exit 1
 fi
 
-# A tag of 'dev' means don't use tag.
-if [ "${GIT_TAG}" = "dev" ]; then
-    GIT_TAG=""
+# Determine if we use official repos with tag or custom branches
+if [ "${BRANCH_PARAM_SPECIFIED}" = "Y" ] || [ "${GIT_TAG}" = "dev" ]; then
+    RELY_ON_TAG="N"
+else
+    RELY_ON_TAG="Y"
+    if [ "${GIT_TAG}" = "last" ]; then
+        echo "Determining the latest tag in AppScale/appscale repo"
+        GIT_TAG=$(curl --fail https://api.github.com/repos/appscale/appscale/tags \
+                  | grep '"name"' | head -1 \
+                  | awk -F ':' '{ print $2 }' | tr --delete ' ,"')
+    fi
+    VERSION="${GIT_TAG}"
 fi
 
 # At this time we expect to be installed in $HOME.
 cd $HOME
 
-# Let's pull the github repositories.
 echo
-echo "Will be using the following github repo:"
-echo "Repo: ${APPSCALE_REPO} Branch: ${APPSCALE_BRANCH}"
-echo "Repo: ${APPSCALE_TOOLS_REPO} Branch: ${APPSCALE_TOOLS_BRANCH}"
-echo "Exit now (ctrl-c) if this is incorrect"
+if [ "${RELY_ON_TAG}" = "Y" ]; then
+    echo "Will be using the following github repos:"
+    echo "AppScale:        ${APPSCALE_REPO} - Tag ${GIT_TAG}"
+    echo "AppScale-Tools:  ${APPSCALE_TOOLS_REPO} - Tag ${GIT_TAG}"
+    if version_ge ${VERSION} 3.7.0; then echo "Cloud-Agents:    ${AGENTS_REPO} - Tag ${GIT_TAG}"; fi
+    if version_ge ${VERSION} 3.8.0; then echo "Thirdparties:    ${THIRDPARTIES_REPO} - Tag ${GIT_TAG}"; fi
```

Comment:
Our development branches firstly get new features and functionally become 3.8 or 4.0, and only during release it actually gets version updated.
So I decreased versions on purpose. Thanks for mentioning that it can cause to redundant clone, but I don't think that cloning these small repos would affect user experience too much.
But I'll have to add all 3.7.0+ tags to agents and thirdparties repos just to keep it working.

Reply:
My understanding was that `VERSION` is from the git tag, in which case development branches seem to be unrelated.

Reply:
`VERSION` is from `--tag` if it is specified and only for showing initial summary message (code lines above this conversation).
`VERSION` is replaced below: `VERSION=$(cat /root/appscale/VERSION | grep -oE "[0-9]+\.[0-9]+\.[0-9]+")`, so version is tested even if tag is not specified (it is taken from `/root/appscale/VERSION`).